client/servers/manager.go

```diff
@@ -0,0 +1,529 @@
+	offline := atomic.LoadInt32(&m.offline)
+	return offline == 1
+}
+
+// FindServer takes out an internal "read lock" and searches through the list
+// of servers to find a "healthy" server.  If the server is actually
+// unhealthy, we rely on Serf to detect this and remove the node from the
+// server list.  If the server at the front of the list has failed or fails
+// during an RPC call, it is rotated to the end of the list.  If there are no
+// servers available, return nil.
+func (m *Manager) FindServer() *Server {
+	l := m.getServerList()
+	numServers := len(l.servers)
+	if numServers == 0 {
+		m.logger.Printf("[WARN] manager: No servers available")
+		return nil
+	}
+
+	// Return whatever is at the front of the list because it is
+	// assumed to be the oldest in the server list (unless -
+	// hypothetically - the server list was rotated right after a
+	// server was added).
+	return l.servers[0]
+}
+
+// getServerList is a convenience method which hides the locking semantics
+// of atomic.Value from the caller.
+func (m *Manager) getServerList() serverList {
+	return m.listValue.Load().(serverList)
+}
+
+// saveServerList is a convenience method which hides the locking semantics
+// of atomic.Value from the caller.
+func (m *Manager) saveServerList(l serverList) {
+	m.listValue.Store(l)
+}
+
+// NumNodes returns the number of approximate nodes in the cluster.
+func (m *Manager) NumNodes() int32 {
+	return atomic.LoadInt32(&m.numNodes)
+}
+
+// SetNumNodes stores the number of approximate nodes in the cluster.
+func (m *Manager) SetNumNodes(n int32) {
+	atomic.StoreInt32(&m.numNodes, n)
+}
+
+// NotifyFailedServer marks the passed in server as "failed" by rotating it
+// to the end of the server list.
+func (m *Manager) NotifyFailedServer(s *Server) {
+	l := m.getServerList()
+
+	// If the server being failed is not the first server on the list,
+	// this is a noop.  If, however, the server is failed and first on
+	// the list, acquire the lock, retest, and take the penalty of moving
+	// the server to the end of the list.
+
+	// Only rotate the server list when there is more than one server
+	if len(l.servers) > 1 && l.servers[0] == s &&
+		// Use atomic.CAS to emulate a TryLock().
+		atomic.CompareAndSwapInt32(&m.notifyFailedBarrier, 0, 1) {
+		defer atomic.StoreInt32(&m.notifyFailedBarrier, 0)
+
+		// Grab a lock, retest, and take the hit of cycling the first
+		// server to the end.
+		m.listLock.Lock()
+		defer m.listLock.Unlock()
+		l = m.getServerList()
+
+		if len(l.servers) > 1 && l.servers[0] == s {
+			l.servers = l.cycleServer()
+			m.saveServerList(l)
+		}
+	}
+}
+
+// NumServers takes out an internal "read lock" and returns the number of
+// servers.  numServers includes both healthy and unhealthy servers.
+func (m *Manager) NumServers() int {
+	l := m.getServerList()
+	return len(l.servers)
+}
+
+// GetServers returns a copy of the current list of servers.
+func (m *Manager) GetServers() []*Server {
```

Comment:
Remove or move to tests where it's used? (See NotifyFailedServer comment)

Reply:
Used by client